### ISSUE TITLE
Not rendering highway ref's with long modifiers

### ISSUE THREAD

**Issue by elijahmathews** [gravitystorm/openstreetmap-carto#4438]
### Expected behavior

Highway ways should have the values of their `ref` tags rendered.

### Actual behavior

Highway ways with `ref` tags that contain long modifiers (e.g. `ref=US 81 Business` or `ref=ND 200 Alternate`) are not having their `ref`'s rendered. They *are* rendered if the modifiers are abbreviated (e.g. `ref=US 90 ALT`), but that probably goes against the [no abbreviations convention](https://wiki.openstreetmap.org/wiki/Naming_conventions#Don.27t_use_abbreviations).

Some abbreviations that are known to be used:
- `ref=... Business` ➞ `ref=... BUS` or `ref=... Bus`
- `ref=... Alternate` ➞ `ref=... ALT` or `ref=... Alt`
- `ref=... Bypass` ➞ `ref=... BYP`
- `ref=... Truck` ➞ `ref=... TRK`

### Links and screenshots illustrating the problem

In Grand Forks, ND, [32nd Avenue South](https://www.openstreetmap.org/way/342349438) is (probably correctly) tagged with `ref=US 81 Business` but has no `ref` rendered, unlike the nearby interstate with `ref=I 29;US 81`:
![ex1](https://user-images.githubusercontent.com/23229627/125148172-0d9e8000-e0f6-11eb-8343-07db6535e192.png)

Meanwhile, in East Bernard, TX, [Main Street](https://www.openstreetmap.org/way/785287636) is (probably incorrectly) tagged with `ref=US 90 ALT` and has its `ref` rendered as expected:
![ex2](https://user-images.githubusercontent.com/23229627/125148220-68d07280-e0f6-11eb-87f0-4bf16c411960.png)

**Comment by imagico** [comment]
Length limit is 11 characters, determined on the longest ref:

https://github.com/gravitystorm/openstreetmap-carto/blob/b86de31da9e44e1a308eb7c8f57c3f4a9bce8956/project.mml#L1735

The concern about this creating bad mapping incentives is valid, i am not sure if the solution can be to maintain a list of abbreviations in all languages of the world (at least not within this style - it could be a valuable idea for a separate project - like also for abbreviations in other labels like street names).  The alternative would be to apply some generic shortening using an ellipsis like suggested for addresses already: https://github.com/gravitystorm/openstreetmap-carto/issues/4160#issuecomment-868987197

**Comment by elijahmathews** [comment]
I suppose even if there were to be some sort of abbreviation lookup table that was used solely for the purpose of rendering highway `ref`'s, that wouldn't necessarily negate bad mapping practices as mappers would presume the displayed labels' format (i.e. a rendered label along the lines "US 81 BUS") is exactly how the `ref` values should be formatted (i.e. `ref=US 81 BUS` rather than `ref=US 81 Business`). The ellipsis (or even just a period) method would probably be better in that case since it would make it clearer there's an abbreviation taking place, although some modifier abbreviations wouldn't work as well in that scheme (e.g. "Truck" is probably better abbreviated as "Trk..." or "Trk." rather than "Tru..." or "Tru.").

To me, perhaps the best scenario would be to add a new tag along the lines of `ref:abbr=*` (and multilingual versions like `ref:es:abbr=*`) to allow mappers to explicitly set the abbreviated version of the `ref`, which renderers would look to first if it's used or else defer to `ref`. However, this would mostly add redundant information to the map, you'd still get some `ref:abbr` values that would probably exceed the 11 character limit (e.g. if a hypothetical but not that unreasonable highway `ref:abbr=ND 1804B ALT` existed), and it necessarily enforces one particular abbreviation scheme.

**Comment by Adamant36** [comment]
> To me, perhaps the best scenario would be to add a new tag along the lines of ref:abbr=*

There's currently the tag `short_name` for abbreviated names. `Short_ref` also exists, but unfortunately it only has 7 uses. That said, it might be worth turning into a proposal and advocating somewhere for wider usage of it.

**Comment by ZeLonewolf** [comment]
Note this would be made redundant if #508 is implemented.

**Comment by imthatleon** [comment]
Just letting you know that this bug is being talked about on [this particular changeset](https://www.openstreetmap.org/changeset/180919370), in regards to [Interstate 80 Business in Sacramento, CA](https://en.wikipedia.org/wiki/Interstate_80_Business_(Sacramento,_California)). Even though it's a business route, it's a major freeway that bisects the city, however the label doesn't show because of this bug and it may cause confusion since there are two I-80s in town. When is a fix coming, and what should be done in the meantime?

**Comment by imagico** [comment]
> When is a fix coming

If and when someone develops a solution for this.  Possible approaches and constrainsts for that have been mentioned above.

> what should be done in the meantime?

A valuable project that would immensely help any kind of map design project that displays name labels is a global abbreviation database for cartographic labels.  Various map design projects have ad hoc abbreviation implementations, cast into framework specific code and limited to select regions and languages.  Like:

https://github.com/cquest/osmfr-cartocss/blob/master/functions.sql
https://github.com/giggls/osml10n/blob/master/lua_osml10/osml10n/street_abbrev.lua

Collecting such information in an implementation independent form with a global scope and in a form that allows contributions without specialized technical knowledge would substantially reduce the work necessary for building high quality OSM based map design w.r.t. labels.

**Comment by dch0ph** [comment]
Issue overlaps with #1513.

**Comment by 1ec5** [comment]
At issue here is a specific subset of abbreviations for a regional route designation practice. Many highway route and recreational trail systems in the United States designate [special routes](https://en.wikipedia.org/wiki/Special_route) that are known by the parent route’s number plus a “banner”.  A few MUTCD-influenced countries have also adopted this practice. In OSM, this banner is appended to the way `ref=*`, appended to the route relation’s `network=*`, and repeated in the route relation’s `modifier=*`.

The repertoire of banners is fixed and not subject to much variation. Any abbreviation of these banners for presentational purposes would not be interchangeable with per-network alphabetic prefixes in route numbers (A, I, US) or with abbreviations of generic words in street names (St., Dr.), even if there’s a match.

[The U.S. federal standard allows](https://mutcd.fhwa.dot.gov/pdfs/11th_Editionr1/mutcd11theditionr1hl.pdf#page=261) for Alternate (Alt.), By-Pass (Byp.), Business (Bus.), Truck, and Temporary (Temp.). OSM Americana also knows of subnational route systems that use Loop, Historic (Hist.), Scenic, Spur, Former (Old.), Toll, Local, and Express (Expr.). Puerto Rico adds Alterna (Alt.), Derivación (Der.), Comercial (Com.), Camión (Cam.), and Provisional (Prov.). Québec adds “Route de contournement” in French (no idea if it can be abbreviated). [Thailand](https://commons.wikimedia.org/wiki/Category:Auxiliary_plates_of_Thailand) has a few in Thai (no abbreviations).

That’s about it. Fortunately, most of these abbreviations are already documented on Wikidata ([example](https://www.wikidata.org/wiki/Q139149664)), so it would be possible to pull out a query in JSON format or whatever you wish.

**Comment by imagico** [comment]
Yes, that was formulated a bit ambiguously - that proper names and refs follow different abbreviation rules should be obvious.

I have not seen any project so far that aims to collect abbreviations for either with a global aim.  Concrete pointers will of course be welcome.

Another problem is, of course, that proper abbreviation requires knowledge of the language.  For `ref` this is even more difficult than for `name`.

But again: Lookup based abbreviation is just one option to address this.

**Comment by 1ec5** [comment]
In this case, isn’t “global” aim sort of like baseball’s “World” Series? Seemingly only a handful of countries designate special routes with banners that can reasonably be abbreviated. (The United States, Canada, Australia… any others?) It may not be strictly necessary to know the language in the way ref, thanks to the limited number of distinct banners and languages. (English, French, and Spanish, and no discordant abbreviations that I know of.) I recognize that this is nothing more than a stroke of luck rather than a fundamentally rigorous approach.

**Comment by imagico** [comment]
> In this case, isn’t “global” aim sort of like baseball’s “World” Series?

More like the United Nations.

There are two possibilities:

* lengthy highway refs that are meant to be abbreviated in labeling are a special particularity, exclusive to a small and strictly limited part of the world.  Then, i am afraid, this is probably out of scope for this style.
* this use of refs is something that mappers agree to be a valid use world wide.  Then any solution we implement to support that should not be limited to those parts of the world that ask most loudly for it.

Or in other words: Mapper consensus on `ref` currently seems to be that it is for locally unique identifiers other than names that are in use by the general public to refer to the feature in question, usually on basis of some classification (like for roads/routes, or for certain types of roads/routes).  That `ref` has to be short is not a requirement.  That we currently only display short refs is owed to the fact that allowing arbitrary length refs for roads is technically difficult and practically most refs world wide are short.  What this issue is about is how we might support longer refs on roads.

What this issue is not about - even if the issue opener uses a case example from the US - is specifically supporting a system to generate short refs from longer refs tagged based in a specific conversion rule set that applies exclusively to a part of the world where less than 10 percent of the world population live.  We could (and ideally probably should) support these rules as part of a larger solution, but not selectively.

**Comment by 1ec5** [comment]
> lengthy highway refs that are meant to be abbreviated in labeling are a special particularity, exclusive to a small and strictly limited part of the world.

Yes, if small is measured in terms of the number of countries or population, it is limited to a small, plainly ignorable portion of the world. (It does however affect an outsized number of roads, mappers, and road users relative to that metric.) For better or worse, OSM Carto creates unrealistic expectations in this portion of the world by rendering any road refs at all. Hypothetically, if this stylesheet were to drop support for way refs in the affected countries or even globally, no one would be clamoring to show “I&nbsp;80 Business” or “I&nbsp;80 Bus” because “I&nbsp;80” wouldn’t show either. (In case you’re wondering, #508 would not be a solution to this issue. The banners still need to be shown somehow.)

> What this issue is about is how we might support longer refs on roads.
> 
> What this issue is not about - even if the issue opener uses a case example from the US - is specifically supporting a system to generate short refs from longer refs tagged based in a specific conversion rule set that applies exclusively to a part of the world where less than 10 percent of the world population live. We could (and ideally probably should) support these rules as part of a larger solution, but not selectively.

Fair enough. Personally I would agree with leaving these words unabbreviated for consistency with the fact that road names remain unabbreviated. The reason this keeps coming up in the U.S. community is that for decades a debate has simmered between mappers who believe these banners should be abbreviated (because refs are more symbolic than names) versus those who believe they should be spelled out (because we generally avoid abbreviations and the banners are generally spelled out on the signs). We adopted the latter many years ago, probably because one of the most common banners, Business, would become “Bus” if abbreviated, potentially misleading users to think it’s a busway (#4226). But a segment of the community has always been dissatisfied with this outcome because the length limit in OSM Carto results in incomplete coverage.

> The concern about this creating bad mapping incentives is valid

If the goal is to avoid misincentivizing mappers to stuff names in way refs, then the 11-character limit is actually detrimental to that goal. We have evidence that mappers are putting [full route names in route refs](https://community.openstreetmap.org/t/veterans-of-foreign-wars-of-the-united-states-memorial-highway-aka-i-83/141703/15) and even way refs _because_ OSM Carto doesn’t render them, to the detriment of data consumers that do otherwise.

**Comment by imagico** [comment]
Ok, i think the various ideas communicated are largely clear.  This issue is open, we definitely would like the arbitrary length limit to be removed - as said: This exists for technical reasons and was never intended to influence mapping in one way or the other. And various possible approaches have been outlined how to accomplish that.  The limit was essentially created under the (wrong) premise that *11 characters ought to be enough for anybody*.

I think the idea with #508 solving this was along the lines that this would involve a fundamental re-design of the shield rendering with specific handling of various systems of design and in that a label string processing would be a natural part.

The whole matter of label abbreviation according to language/culture specific rules is probably something that would be best to discuss independent of specific use cases like this first.  Names and `ref` are not the only cases where this is a concern and a broader perspective can surely help.  Likewise the risk of mappers abbreviating arbitrarily and inconsistently for maps that flatly render what is tagged is a larger concern.
